CoroutineDescendantAxis.kt revamp.

xDido opened the pull request.

xDido committed:
CoroutineDescendantAxis.kt revamp.

JohannesLichtenberger commented:
Hey, thanks a lot, but I think we need to first check if the right sibling is on the same page (meaning it'll be cached already). Only if it's not we can use another trx to fetch it in the meantime.

In general I'm not sure if synchronization overhead might be way too much, I don't really know, but I guess it might be.

JohannesLichtenberger commented:
Oh, saw that you did this, but you can use `PageReadOnlyTrx::pageKey` to get the page number of a node.

JohannesLichtenberger commented:
BTW: the integration tests fail

xDido committed:
Merge branch 'sirixdb:main' into coroutine-descendant-axis-fix

xDido committed:
I replaced a simplified page key calculation in CoroutineDescendantAxis's shouldParallelizeSubtree method with an accurate calculation using
PageReadOnlyTrx.pageKey(). still tests fail because of NPE regarding the atomicBoolean initialization.

JohannesLichtenberger commented:
You also most likely get the descendantCount of nodes directly. Furthermore, you can navigate with `moveToFirstChild()`, `moveToRightSibling()`. Waiting 1s for a parallel result is also way too long and if there's nothing in the queue there's nothing. Also, I think you wouldn't use a BlockingQueue in Kotlin. You typically use a `Channel` instead.

I don't know why your logger seems to be null, but maybe you're missing a configuration?

Over here the precondition fails? https://github.com/sirixdb/sirix/blob/5b8dd63137e55ffe20358f93512edd81ef67f973/bundles/sirix-core/src/main/java/io/sirix/utils/LogWrapper.java#L45

xDido committed:
refactor: use Kotlin Channel and proper cursor navigation in CoroutineDescendantAxis

Replace blocking queue with Channel, use moveToFirstChild()/moveToRightSibling()
methods, and access descendantCount directly for better performance.

xDido committed:
add logger.trace to the LogWrapper

xDido committed:
updated logger constructor with a logging message.

xDido committed:
added logging to the code.

xDido commented:
I have added a logging message in the LogWrapper constructor as well as throughout the entire class. I ran the tests, and nothing in the stack trace indicates any issues originating from the log wrapper. It simply states that it's null on the first call of log.debug at line 105 in the reset function.  
![image](https://github.com/user-attachments/assets/66d9a55b-06e0-4671-9390-6a35742f3132)
